lib/nixos-system.nix

```diff
@@ -1,43 +1,93 @@
 { self, lib, home-manager }:
 
-{ modules ? [ ], globalSpecialArgs ? { }, nixosSpecialArgs ? { }, hmSpecialArgs ? { }, ... } @ args:
-lib.nixosSystem (lib.recursiveUpdate (removeAttrs args [ "globalSpecialArgs" "nixosSpecialArgs" "hmSpecialArgs" ]) {
+with lib;
+
+{
+  # The global modules are included in both NixOS and home-manager.
+  globalModules ? [ ]
+
+  # Home-manager specific modules.
+, hmModules ? [ ]
+
+  # NixOS specific modules.
+, nixosModules ? [ ]
+
+  # The global extra arguments are included in both NixOS and home-manager.
+, globalSpecialArgs ? { }
+
+  # Home-manager specific extra arguments.
+, hmSpecialArgs ? { }
+
+  # NixOS specific extra arguments.
+, nixosSpecialArgs ? { }
+
+, ...
+} @ args:
+let
+  args' = removeAttrs args [
+    "globalModules"
+    "nixosModules"
+    "hmModules"
```

Comment:
The ordering of those is not the same as above. (Sorry for nitpicking)

Reply:
Done